sklearn/utils/__init__.py

```diff
@@ -1182,3 +1204,39 @@ def is_abstract(c):
     # itemgetter is used to ensure the sort does not extend to the 2nd item of
     # the tuple
     return sorted(set(estimators), key=itemgetter(0))
+
+
+def _safe_tags(estimator, key=None, default=None):
+    """Safely get estimator tags for common checks.
+
+    :class:`~sklearn.BaseEstimator` provides the estimator tags machinery.
+    However, if a compatible estimator does not inherit from this base class,
+    we should default to the default tag.
+
+    Parameters
+    ----------
+    estimator : estimator object
+        The estimator from which to get the tag.
+    key : str, default=None
+        Tag name to get. By default (`None`), all tags are returned.
+    default : list of {str, dtype} or bool, default=None
+        When `key is not None`, if the tag was not set in the estimator, the
+        default value set in `sklearn.utils._DEFAULT_TAGS` will be returned.
+        `default` allows to overwrite the default value.
+
+    Returns
+    -------
+    tags : dict
+        The estimator tags.
+    """
+    if hasattr(estimator, "_get_tags"):
+        if key is not None:
+            default = _DEFAULT_TAGS[key] if default is None else default
```

Comment:
If `_get_tags` exists, I think it's fair to assume that it's `BaseEstimator._get_tags`, and thus passing a default makes no sense because `_DEFAULT_TAGS` will always be used (in `_get_tags`).

Or said differently, if `_get_tags` exists, then all keys should exist.

So I'd raise a ValueError if `_get_tags` exists and `default` is not None.

Reply:
I would agree but it was breaking our test because we did not make that assumption in the test themselves:

https://github.com/scikit-learn/scikit-learn/blob/9d394c2daab6df104cef115ebd69f802cc327347/sklearn/feature_selection/tests/test_rfe.py#L59-L60

So here we have 2 choices:

* assume that when `_get_tags` they should have default key and we raise a `ValueError` instead. We should as well document it (probably in the estimator tags section);
* be permissive.

Reply:
`cuml` is for instance planning to implement their own `_get_tags` since they have their own base class. So we are not quite precise in the documentation that the `_DEFAULT_TAGS` should be implemented for sure.

I modify the function such that it handles the case that default tags are not implemented and overwritten but I don't have issues to be stricter depending on opinions of everyone on that matter.

Reply:
IMHO we should update the docs to explicitly say that all keys should exist if `_get_tags` exists. It's private, after all.

(and we should update the test you mentioned)

Reply:
@thomasjpfan Since that you are around. Do you have an opinion on this matter?

Reply:
I implemented what @NicolasHug suggested. I added the specificity in the documentation and a common check for when an estimator is implementing `_get_tags`.